lib/core/src/persist/sync.rs

```diff
@@ -0,0 +1,239 @@
+        Ok(sync_state)
+    }
+
+    pub(crate) fn set_sync_state(&self, sync_state: SyncState) -> Result<()> {
+        let con = self.get_connection()?;
+
+        con.execute(
+            "
+            INSERT OR REPLACE INTO sync_state(data_id, record_id, record_revision)
+            VALUES (:data_id, :record_id, :record_revision)
+        ",
+            named_params! {
+                ":data_id": &sync_state.data_id,
+                ":record_id": &sync_state.record_id,
+                ":record_revision": &sync_state.record_revision,
+            },
+        )?;
+
+        Ok(())
+    }
+
+    pub(crate) fn get_sync_settings(&self) -> Result<Option<SyncSettings>> {
+        let con = self.get_connection()?;
+
+        let sync_settings = con.query_row(
+            "
+            SELECT 
+                remote_url,
+                latest_revision
+            FROM sync_settings
+            WHERE id = 1
+        ",
+            [],
+            |row| {
+                Ok(match row.get(0)? {
+                    Some(remote_url) => Some(SyncSettings {
+                        remote_url,
+                        latest_revision: row.get(1)?,
+                    }),
+                    None => None,
+                })
+            },
+        )?;
+
+        Ok(sync_settings)
+    }
+
+    pub(crate) fn set_sync_settings(&self, sync_settings: SyncSettings) -> Result<()> {
+        let con = self.get_connection()?;
+
+        con.execute(
+            "
+            INSERT OR REPLACE INTO sync_settings(data_id, remote_url, latest_revision)
+            VALUES (1, :remote_url, :latest_revision)
+        ",
+            named_params! {
+                ":remote_url": &sync_settings.remote_url,
+                ":latest_revision": &sync_settings.latest_revision,
+            },
+        )?;
+
+        Ok(())
+    }
+
+    pub(crate) fn get_incoming_record(&self, record_id: &str) -> Result<Option<Record>> {
+        let con = self.get_connection()?;
+
+        let record = con.query_row(
+            "
+            SELECT 
+                revision,
+                id,
+                schema_version,
+                data
+            FROM sync_incoming
+            WHERE id = :id
+        ",
+            named_params! {
+                ":id": record_id,
+            },
+            |row| {
+                Ok(match row.get(0)? {
+                    Some(revision) => Some(Record {
+                        revision,
+                        id: row.get(1)?,
+                        schema_version: row.get(2)?,
+                        data: row.get(3)?,
+                    }),
+                    None => None,
+                })
+            },
+        )?;
+
+        Ok(record)
+    }
+
+    pub(crate) fn set_incoming_record(&self, record: &Record) -> Result<()> {
+        let con = self.get_connection()?;
+
+        con.execute(
+            "
+            INSERT INTO sync_incoming(revision, id, schema_version, data)
```

Comment:
I think we should use here insert or replace so existing incoming record will be overriden and we won't attempt to add more incoming for the same record id.